### PR TITLE
Nccl gloo refit for RL

### PR DESCRIPTION
**Pull request by wdykas**
# What does this PR do ?
This is separating out just the GLOO and NCCL refit methods from https://github.com/NVIDIA/Megatron-LM/pull/2696 for easier merging.

This allows rl training with different inference and training parallelisms. The fast path NVshmem backend will be added after this goes in.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @megatron-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @megatron-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by wdykas** [comment]
/ok to test [3de5e9a](https://github.com/NVIDIA/Megatron-LM/pull/2812/commits/3de5e9a9977565da638a738926efe7a110d27e6f)

**Comment by wdykas** [comment]
@kanz-nv Please review this is the nccl and gloo separated out.

**Comment by wdykas** [comment]
@santhnm2  @sidsingh-nvidia This is just the nccl and gloo refit stuff separated out so should be simpler to review. We will do the nvshmem separately.

File: megatron/core/resharding/utils.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    for part in param_name.split('.'):
+        if (
+            part.startswith('weight')
+            and len(part) > len('weight')
+            and part[len('weight') :].isdigit()
+        ):
+            return int(part[len('weight') :])
+        if part.startswith('bias') and len(part) > len('bias') and part[len('bias') :].isdigit():
+            return int(part[len('bias') :])
+    return None
```
Comment (santhnm2) [inline megatron/core/resharding/utils.py:123]: Have we tested this with Nanov3?

Comment (wdykas) [reply megatron/core/resharding/utils.py:123]: I have tested with Te GroupedGemm moe implementation but now Nanov3. Do we have a Nanov3 model in RL yet?

Comment (tdene) [reply megatron/core/resharding/utils.py:123]: See #2742 

@jon-barker is just adding further testing to that PR before merging it, but other than that it works.

Comment (wdykas) [reply megatron/core/resharding/utils.py:123]: ok this has been tested outside of RL for MoE resharding in the unit tests. For end2end RL only dense has been tested as that is what we have right now on main

File: megatron/rl/rl_utils.py [hunk 2]
```diff
@@ -81,6 +93,121 @@
...
+    """
+    args = get_args()
+    if not args.rl_offload_inference_model_weights_when_idle:
+        return
+    if args.rl_inference_model_unified_memory_level != 1:
+        return
+
+    device = -1 if to_cpu else int(torch.cuda.current_device())
+    advise_managed_module_parameters_preferred_location(model_core, device=device, include_buffers=True)
+    nbytes = prefetch_managed_module_parameters(model_core, device=device, include_buffers=True)
```
Comment (wdykas) [inline megatron/rl/rl_utils.py:112]: We can maybe do this on a side stream.  I leave it for now since I think in our current case I dont think we care

File: megatron/core/inference/unified_memory.py [hunk 3]
```diff
@@ -160,14 +215,16 @@ def compile_allocator():
                 _cpa = CUDAPluggableAllocator(_so_path, "managed_malloc", "managed_free")
                 _alloc = _cpa.allocator()
                 _compilation_state = CompilationState.SUCCESS
+                _compilation_error = None
         except (RuntimeError, ImportError, OSError, UnifiedMemoryCompileTimeoutError) as e:
+            _compilation_error = str(e)
             warnings.warn(f"Failed to create unified memory mempool: '{e}'.")
```
Comment (Skylion007) [inline megatron/core/inference/unified_memory.py:221]: Specify the stacklevel you want here.

**Review by shanmugamr1992 — APPROVED** [review]

File: megatron/core/resharding/utils.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    if len(grouped_by_dp) == 1:
+        return src_meta_list[0]
+
+    # Determine this destination rank's index within its DP group (if any).
+    dst_dp_ranks = dst_metadata.data_parallel_group_ranks or []
+    if dst_dp_ranks and dst_rank in dst_dp_ranks:
+        dst_dp_index = dst_dp_ranks.index(dst_rank)
+    else:
+        # Fallback: treat as the first DP index.
+        dst_dp_index = 0
```
Comment (tdene) [inline megatron/core/resharding/utils.py:263]: When would this fallback trigger?

File: megatron/core/resharding/copy_services/nccl_copy_service.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+                    )
+                with torch.no_grad():
+                    with torch.cuda.stream(self._copy_stream):
+                        recv_op.tensor.copy_(send_op.tensor)
+
+        p2p_ops = []
+        for op in remote_sends:
+            p2p_ops.append(dist.P2POp(dist.isend, op.tensor, op.dest_rank))
+        for op in remote_recvs:
+            p2p_ops.append(dist.P2POp(dist.irecv, op.tensor, op.src_rank))
```
Comment (tdene) [inline megatron/core/resharding/copy_services/nccl_copy_service.py:114]: I haven't been able to figure this out on my own. Note that this applies to both NCCL and GLOO.

P2P ops have to be in matching order. `SEND -> SEND -> RECV` on rank X has to be matched by `RECV -> RECV -> SEND` on rank Y. See torch's docs:

> The order of the isend/irecv in the list matters and it needs to match with corresponding isend/irecv on the remote end.

It seems to me that all ranks list SEND first and RECV after. That would be undefined behavior. What am I missing?

File: megatron/rl/rl_utils.py [hunk 6]
```diff
@@ -331,14 +458,33 @@ def get_environment_rollouts(
...
 
+    # If we have seperate training and inference models we to refit weights from the training model to the inference model.
+    if inference_model is not None:
+        # If the separate inference model weights were prefetched to CPU while idle, bring them
+        # back to GPU before refit/copy and before any CUDA-graph'd inference.
+        with nvtx_range("prefetch-inference-model-weights-to-gpu"):
+            inf_core = unwrap_model(inference_model[0])
+            _maybe_prefetch_separate_inference_model_weights(inf_core, to_cpu=False)
+        swap_model_weights(model, inference_model, args.refit_method)
+        if args.rl_verify_model_weights_swap and args.curr_iteration == 0:
```
Comment (tdene) [inline megatron/rl/rl_utils.py]: Since this is a debug flag anyway, I don't agree with only doing it on iteration 0 instead of doing it on every iteration. I don't envision this flag being turned on during real runs, so no need to prevent perf hits.

File: megatron/core/resharding/copy_services/nccl_copy_service.py (hunk 5 above)
Comment (wdykas) [reply megatron/core/resharding/copy_services/nccl_copy_service.py:114]: https://docs.pytorch.org/docs/stable/distributed.html#torch.distributed.batch_isend_irecv This is following the example provided by the documentation. I am told this is valid ordering by the pytorch team

File: megatron/training/training.py [hunk 7]
```diff
@@ -721,6 +729,103 @@ def pretrain(
     print_datetime('after model, optimizer, and learning rate ' 'scheduler are built')
     config = get_model_config(model[0])
 
+    # Build a separate inference model for RL if requested.
+    inference_model = None
+    if args.perform_rl_step:
+        pg_collection = None
+        if args.rl_inference_tensor_model_parallel_size is not None:
```
Comment (wdykas) [inline megatron/training/training.py:726]: put it into its own function for parallel state creation

File: megatron/training/training.py [hunk 8]
```diff
@@ -721,6 +729,103 @@ def pretrain(
...
+                    model_provider,
+                    model_type,
+                    wrap_with_ddp=False,
+                    pg_collection=inference_pg_collection,
+                    config=inference_config,
+                )
+            inference_model[0].eval()
+
+            # If requested, immediately prefetch weights to CPU to keep them off GPU when idle.
+            if (
```
Comment (wdykas) [inline megatron/training/training.py]: remove this from here because we want them on gpu right away

File: megatron/core/resharding/utils.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+
+    # Which rank owns this param
+    owner_rank: int = -1
+
+    tensor_parallel_group_ranks: list[int] | None = None
+    expert_parallel_group_ranks: list[int] | None = None
+    data_parallel_group_ranks: list[int] | None = None
+    pipeline_parallel_group_ranks: list[int] | None = None
+
+    # Canonicalization for EP per-expert params
```
Comment (wdykas) [inline megatron/core/resharding/utils.py]: Make this more clear what this means

File: megatron/core/resharding/copy_services/nccl_copy_service.py (hunk 5 above)
Comment (wdykas) [reply megatron/core/resharding/copy_services/nccl_copy_service.py:114]: talked offline about this

File: megatron/rl/rl_utils.py (hunk 6 above)
Comment (wdykas) [reply megatron/rl/rl_utils.py]: done

**Comment by wdykas** [comment]
/ok to test [679b22b](https://github.com/NVIDIA/Megatron-LM/pull/2812/commits/679b22bbf9161b70fed431b5ae8f4394184a7727)

**Comment by wdykas** [comment]
/ok to test [0980a61](https://github.com/NVIDIA/Megatron-LM/pull/2812/commits/0980a6177a32c90560fe78663c33b7c991fc6cef)

**Comment by wdykas** [comment]
/ok to test [0980a61](https://github.com/NVIDIA/Megatron-LM/pull/2812/commits/0980a6177a32c90560fe78663c33b7c991fc6cef)

**Comment by wdykas** [comment]
/ok to test [0980a61](https://github.com/NVIDIA/Megatron-LM/pull/2812/commits/0980a6177a32c90560fe78663c33b7c991fc6cef)

**Comment by wdykas** [comment]
/ok to test [98c0d2e](https://github.com/NVIDIA/Megatron-LM/pull/2812/commits/98c0d2ee999367c3ea2406991b92c93a3e2716d4)

File: megatron/core/resharding/planner.py [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+
+    # PP filtering (simple, symmetric)
+    src_pp_ranks = src_metadata.pipeline_parallel_group_ranks
+    dst_pp_ranks = dst_metadata.pipeline_parallel_group_ranks
+    if len(dst_pp_ranks) > 1 and my_global_rank not in dst_pp_ranks:
+        return []
+    if len(src_pp_ranks) > 1 and len(dst_pp_ranks) > 1:
+        my_dst_pp_rank = _get_rank_in_group(my_global_rank, dst_pp_ranks)
+        if my_dst_pp_rank >= len(src_pp_ranks):
+            return []
```
Comment (kvareddy) [inline megatron/core/resharding/planner.py]: Does this mean we don't support resharding across different pipeline parallelism conifgs?

File: megatron/core/resharding/planner.py [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+            return descriptors
+        src_stride = src_metadata.partition_stride if src_metadata.is_tp else 1
+        dst_stride = dst_metadata.partition_stride if dst_metadata.is_tp else 1
+
+        # Size conservation check on partition dim
+        src_world = len(src_tp_ranks)
+        dst_world = len(dst_tp_ranks)
+        src_local = src_metadata.shape[tp_dim]
+        dst_local = dst_metadata.shape[tp_dim]
+        if src_world * src_local != dst_world * dst_local:
```
Comment (kvareddy) [inline megatron/core/resharding/planner.py:48]: Shouldn't this be an assert?

File: megatron/core/resharding/planner.py [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+    descriptors = _build_descriptors_for_param(src_metadata=src_metadata, dst_metadata=dst_metadata)
+    if descriptors:
+        return _plan_multi_dim_lcm(
+            param_name=param_name,
+            src_metadata=src_metadata,
+            dst_metadata=dst_metadata,
+            descriptors=descriptors,
+            my_global_rank=my_global_rank,
+        )
+    # DP / replicated fallback
```
Comment (kvareddy) [inline megatron/core/resharding/planner.py:205]: DP's fallback is only applicable when TP/EP are not present, right?

File: megatron/core/resharding/planner.py [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+        return []
+    if len(src_pp_ranks) > 1 and len(dst_pp_ranks) > 1:
+        my_dst_pp_rank = _get_rank_in_group(my_global_rank, dst_pp_ranks)
+        if my_dst_pp_rank >= len(src_pp_ranks):
+            return []
+
+    # Regular TP/DP planning with EP-resolved metadata
+    descriptors = _build_descriptors_for_param(src_metadata=src_metadata, dst_metadata=dst_metadata)
+    if descriptors:
+        return _plan_multi_dim_lcm(
```
Comment (kvareddy) [inline megatron/core/resharding/planner.py:198]: Where do we do TP+DP together?

File: megatron/core/resharding/planner.py (hunk 12 above)
Comment (wdykas) [reply megatron/core/resharding/planner.py:205]: yes

File: megatron/core/resharding/planner.py (hunk 13 above)
Comment (wdykas) [reply megatron/core/resharding/planner.py:198]: The DP is done by choosing the DP replica that we are using first.  The code first chooses which DP replica to grab from for the destination, then the LCM decides how to do TP

File: megatron/core/resharding/planner.py (hunk 11 above)
Comment (wdykas) [reply megatron/core/resharding/planner.py:48]: hmm maybe, I am trying to think about the case for why I had this here previously.  I may be left over from some merges that got messed up.

Comment (wdykas) [reply megatron/core/resharding/planner.py:48]: OK I remember the purpose of this used to be to catch params that were marked as TP but not actually tensor parallel. Essentially DP replicated params like norms. However this is no longer needed after marking the params properly with TP meta data. I have raised an assertion there is another in the LCM planner as well.

File: megatron/core/resharding/planner.py (hunk 10 above)
Comment (wdykas) [reply megatron/core/resharding/planner.py]: yes this is dumb of me...
I removed this and added a test case to the unit tests for TP1,PP2 -> TP2,PP4 that passes

**Comment by wdykas** [comment]
/ok to test [63bdab5](https://github.com/NVIDIA/Megatron-LM/pull/2812/commits/63bdab5267bfb57b00b041635a4cd79bb9c9d46b)

File: megatron/core/resharding/planner.py [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+    return ops
+
+
+def _plan_dp_recv(
+    param_name: str,
+    src_metadata: ParameterMetadata,
+    dst_metadata: ParameterMetadata,
+    my_global_rank: int,
+) -> list[tuple[int, tuple[slice, ...], tuple[slice, ...]]]:
+    """Plan DP transfer for a replicated (non-TP) parameter (receiver side)."""
```
Comment (wdykas) [inline megatron/core/resharding/planner.py]: add comments on the DP, also maybe rename

File: megatron/core/resharding/planner.py (hunk 10 above)
Comment (wdykas) [reply megatron/core/resharding/planner.py]: add description about how pipelining is handled

**Comment by wdykas** [comment]
/ok to test [9a72692](https://github.com/NVIDIA/Megatron-LM/pull/2812/commits/9a72692078dec0a349a6b3860cd38a32953e6ce7)
